cpp/examples/tpch/q5.cpp

```diff
@@ -0,0 +1,190 @@
+    auto one_minus_disc = cudf::binary_operation(one, disc, cudf::binary_operator::SUB, disc.type());
+    auto extended_price = table->column("l_extendedprice");
+
+    auto disc_price_type = cudf::data_type{cudf::type_id::DECIMAL64, -4};
+    auto disc_price = cudf::binary_operation(extended_price, one_minus_disc->view(), cudf::binary_operator::MUL, disc_price_type);
+    return disc_price;
+}
+
+int main(int argc, char const** argv) {
+    auto args = parse_args(argc, argv);
+
+    // Use a memory pool
+    auto resource = create_memory_resource(args.use_memory_pool);
+    rmm::mr::set_current_device_resource(resource.get());
+
+    Timer timer;
+
+    // 1. Read out the tables from parquet files
+    //    while pushing down column projections and filter predicates
+    auto o_orderdate_ref = cudf::ast::column_reference(2);
+    
+    auto o_orderdate_lower = cudf::timestamp_scalar<cudf::timestamp_D>(days_since_epoch(1994, 1, 1), true);
+    auto o_orderdate_lower_limit = cudf::ast::literal(o_orderdate_lower);
+    auto o_orderdate_pred_a = cudf::ast::operation(
+        cudf::ast::ast_operator::GREATER_EQUAL,
+        o_orderdate_ref,
+        o_orderdate_lower_limit
+    );
+    
+    auto o_orderdate_upper = cudf::timestamp_scalar<cudf::timestamp_D>(days_since_epoch(1995, 1, 1), true);
+    auto o_orderdate_upper_limit = cudf::ast::literal(o_orderdate_upper);
+    auto o_orderdate_pred_b = cudf::ast::operation(
+        cudf::ast::ast_operator::LESS,
+        o_orderdate_ref,
+        o_orderdate_upper_limit
+    );
+
+    auto o_orderdate_pred = std::make_unique<cudf::ast::operation>(
+        cudf::ast::ast_operator::LOGICAL_AND,
+        o_orderdate_pred_a,
+        o_orderdate_pred_b
+    );
+
+    auto r_name_ref = cudf::ast::column_reference(1); 
+    auto r_name_value = cudf::string_scalar("ASIA");
+    auto r_name_literal = cudf::ast::literal(r_name_value);
+    auto r_name_pred = std::make_unique<cudf::ast::operation>(
+        cudf::ast::ast_operator::EQUAL,
+        r_name_ref,
+        r_name_literal
+    );
+
+    auto customer = read_parquet(
+        args.dataset_dir + "customer/part-0.parquet", {"c_custkey", "c_nationkey"});
+    auto orders = read_parquet(
+        args.dataset_dir + "orders/part-0.parquet", 
+        {"o_custkey", "o_orderkey", "o_orderdate"},
+        std::move(o_orderdate_pred)
+    );
+    auto lineitem = read_parquet(
+        args.dataset_dir + "lineitem/part-0.parquet", 
+        {"l_orderkey", "l_suppkey", "l_extendedprice", "l_discount"}
+    );
+    auto supplier = read_parquet(
+        args.dataset_dir + "supplier/part-0.parquet", {"s_suppkey", "s_nationkey"}
+    );
+    auto nation = read_parquet(
+        args.dataset_dir + "nation/part-0.parquet", {"n_nationkey", "n_regionkey", "n_name"}
+    );
+    auto region = read_parquet(
+        args.dataset_dir + "region/part-0.parquet", 
+        {"r_regionkey", "r_name"},
+        std::move(r_name_pred)
+    );
+
+    // 2. Perform the joins
```

Comment:
Do we need to explain how we chose this order for the joins?

Reply:
Maybe we can link to the reference implementations that we used ?

Reply:
Fixed the join order for Q9. All join orders according to DuckDB's query plan

Reply:
Please document this in the comments as well.